File: arrays_and_strings/reverse_string.rb
# @param {String} s
# @return {String}
def reverse_string(s)
  return s if s.empty?
  0.upto(s.size/2 - 1) do |i|
    s[i], s[-i -1] = s[-i - 1], s[i]
  end

  s
end


require 'test/unit'
extend Test::Unit::Assertions

assert_equal(reverse_string('a.'), '.a')
assert_equal(reverse_string('hello'), 'olleh')
assert_equal(reverse_string("a ba"), "ab a")
assert_equal(reverse_string("abcd"), "dcba")

File: strings/reverse_string.py
class Solution(object):
    def reverseString(self, s):
        l = len(s)
        if l < 2:
            return s
        return self.reverseString(s[l/2:]) + self.reverseString(s[:l/2])


class SolutionClassic(object):
    def reverseString(self, s):
        r = list(s)
        i, j  = 0, len(r) - 1
        while i < j:
            r[i], r[j] = r[j], r[i]
            i += 1
            j -= 1

        return "".join(r)

class SolutionPythonic(object):
    def reverseString(self, s):
        return s[::-1]

# 344. Reverse String
# https://leetcode.com/problems/reverse-string/description/

# Time: O(n)
# Space: O(1)

# require 'test/unit'
# extend Test::Unit::Assertions

# assert_equal(reverse_string("reverse"), "esrever")
# assert_equal(reverse_string("abcd"), "dcba")
# assert_equal(reverse_string("cdc"), "cdc")
